File: shared/src/commonTest/kotlin/com/cafesito/shared/domain/brew/BrewEngineTest.kt
package com.cafesito.shared.domain.brew

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class BrewEngineTest {

    @Test
    fun caffeine_does_not_change_with_cup_size_if_grams_are_equal() {
        val small = BrewEngine.estimateCaffeineMg(
            BrewCaffeineInput(
                source = BrewSource.DIARY,
                methodOrPreparation = "Espresso",
                coffeeGrams = 18.0,
                hasCaffeine = true
            )
        )
        val large = BrewEngine.estimateCaffeineMg(
            BrewCaffeineInput(
                source = BrewSource.DIARY,
                methodOrPreparation = "Espresso",
                coffeeGrams = 18.0,
                hasCaffeine = true
            )
        )
        assertEquals(small, large)
    }

    @Test
    fun decaf_is_lower_than_regular_for_same_grams() {
        val regular = BrewEngine.estimateCaffeineMg(
            BrewCaffeineInput(BrewSource.BREWLAB, "V60", 15.0, hasCaffeine = true)
        )
        val decaf = BrewEngine.estimateCaffeineMg(
            BrewCaffeineInput(BrewSource.BREWLAB, "V60", 15.0, hasCaffeine = false)
        )
        assertTrue(decaf < regular)
    }

    @Test
    fun default_pourover_timeline_has_three_phases() {
        val phases = BrewEngine.timelineForMethod("Hario V60", 300)
        assertEquals(3, phases.size)
        assertEquals("Bloom", phases[0].label)
    }

    @Test
    fun contract_vectors_match_expected_outputs() {
        assertEquals(
            113,
            BrewEngine.estimateCaffeineMg(
                BrewCaffeineInput(BrewSource.DIARY, "Espresso", 18.0, hasCaffeine = true)
            )
        )
        assertEquals(
            91,
            BrewEngine.estimateCaffeineMg(
                BrewCaffeineInput(BrewSource.BREWLAB, "Hario V60", 15.0, hasCaffeine = true)
            )
        )
        assertEquals(
            6,
            BrewEngine.estimateCaffeineMg(
                BrewCaffeineInput(BrewSource.BREWLAB, "Hario V60", 15.0, hasCaffeine = false)
            )
        )
        assertEquals(
            125,
            BrewEngine.estimateCaffeineMg(
                BrewCaffeineInput(BrewSource.DIARY, "Moca", 18.0, hasCaffeine = true)
            )
        )
    }

    @Test
    fun espresso_timeline_contract_vector() {
        val phases = BrewEngine.timelineForMethod("Espresso", 36)
        assertEquals(1, phases.size)
        assertEquals("Extraccion", phases[0].label)
    }

    @Test
    fun cup_size_label_uses_nearest_volume() {
        assertEquals("Espresso", BrewEngine.cupSizeLabelForAmountMl(36))
        assertEquals("Mediano", BrewEngine.cupSizeLabelForAmountMl(260))
        assertEquals("Grande", BrewEngine.cupSizeLabelForAmountMl(410))
    }

    @Test
    fun rapido_caffeine_approximate_by_type_and_size() {
        val espresso30 = BrewEngine.estimateCaffeineMg(
            BrewCaffeineInput(BrewSource.BREWLAB, BREW_METHOD_OTROS, 0.0, hasCaffeine = true, amountMl = 30, drinkType = "Espresso")
        )
        assertTrue(espresso30 in 60..70)
        val mediano = BrewEngine.estimateCaffeineMg(
            BrewCaffeineInput(BrewSource.BREWLAB, BREW_METHOD_OTROS, 0.0, hasCaffeine = true, amountMl = 275, drinkType = "Americano")
        )
        assertTrue(mediano in 140..150)
        val decaf = BrewEngine.estimateCaffeineMg(
            BrewCaffeineInput(BrewSource.BREWLAB, BREW_METHOD_OTROS, 0.0, hasCaffeine = false, amountMl = 180, drinkType = "Filter")
        )
        assertTrue(decaf in 0..10)
    }

    @Test
    fun rapido_approximate_coffee_grams_by_type_and_size() {
        assertEquals(18, BrewEngine.approximateCoffeeGramsForRapido(30, "Espresso"))
        assertTrue(BrewEngine.approximateCoffeeGramsForRapido(60, "Espresso") in 35..37)
        assertTrue(BrewEngine.approximateCoffeeGramsForRapido(275, "Americano") in 15..17)
        assertTrue(BrewEngine.approximateCoffeeGramsForRapido(180, "Filter") in 11..13)
    }

    @Test
    fun method_profile_contract_vector_for_espresso() {
        val profile = BrewEngine.methodProfileFor("Espresso")
        assertEquals(25, profile.waterMinMl)
        assertEquals(60, profile.waterMaxMl)
        assertEquals(2.0, profile.defaultRatio)
    }

    @Test
    fun brewing_process_advice_is_dynamic_by_phase_and_time() {
        val advice = BrewEngine.brewingProcessAdvice(
            method = "Hario V60",
            ratio = 16.0,
            waterMl = 300,
            phaseLabel = "Bloom",
            remainingInPhaseSeconds = 4
        )
        assertTrue(advice.contains("Asegura saturacion completa"))
        assertTrue(advice.contains("4 s"))
    }
}

File: webApp/src/core/brewEngine.test.ts
import { describe, expect, it } from "vitest";
import { approximateCoffeeGramsForRapido, cupSizeLabelForAmountMl, estimateCaffeineMg, hasCaffeineFromLabel } from "./brewEngine";
import { getBrewingProcessAdvice, getBrewTimelineForMethod } from "./brew";
import { normalizeLookupText } from "./text";

describe("brewEngine", () => {
  it("keeps caffeine total stable when grams are equal", () => {
    const small = estimateCaffeineMg({
      source: "diary",
      methodOrPreparation: "Espresso",
      coffeeGrams: 18,
      hasCaffeine: true
    });
    const large = estimateCaffeineMg({
      source: "diary",
      methodOrPreparation: "Espresso",
      coffeeGrams: 18,
      hasCaffeine: true
    });
    expect(large).toBe(small);
  });

  it("detects decaf variants from label", () => {
    expect(hasCaffeineFromLabel("No")).toBe(false);
    expect(hasCaffeineFromLabel("Sin cafeina")).toBe(false);
    expect(hasCaffeineFromLabel("Si")).toBe(true);
  });

  it("matches contract vectors for caffeine outputs", () => {
    expect(
      estimateCaffeineMg({
        source: "diary",
        methodOrPreparation: "Espresso",
        coffeeGrams: 18,
        hasCaffeine: true
      })
    ).toBe(113);

    expect(
      estimateCaffeineMg({
        source: "brewlab",
        methodOrPreparation: "Hario V60",
        coffeeGrams: 15,
        hasCaffeine: true
      })
    ).toBe(91);

    expect(
      estimateCaffeineMg({
        source: "brewlab",
        methodOrPreparation: "Hario V60",
        coffeeGrams: 15,
        hasCaffeine: false
      })
    ).toBe(6);

    expect(
      estimateCaffeineMg({
        source: "diary",
        methodOrPreparation: "Moca",
        coffeeGrams: 18,
        hasCaffeine: true
      })
    ).toBe(125);
  });

  it("matches contract vectors for timeline outputs", () => {
    const pourover = getBrewTimelineForMethod("Hario V60", 300);
    expect(pourover).toHaveLength(3);
    expect(pourover[0]?.label).toBe("Bloom");

    const espresso = getBrewTimelineForMethod("Espresso", 36);
    expect(espresso).toHaveLength(1);
    expect(normalizeLookupText(espresso[0]?.label ?? "")).toBe("extraccion");
  });

  it("maps cup size by nearest volume", () => {
    expect(cupSizeLabelForAmountMl(36)).toBe("Espresso");
    expect(cupSizeLabelForAmountMl(260)).toBe("Mediano");
    expect(cupSizeLabelForAmountMl(410)).toBe("Grande");
  });

  it("Rápido: approximate coffee grams by type and size", () => {
    expect(approximateCoffeeGramsForRapido(30, "Espresso")).toBe(18);
    expect(approximateCoffeeGramsForRapido(60, "Espresso")).toBeGreaterThanOrEqual(35);
    expect(approximateCoffeeGramsForRapido(60, "Espresso")).toBeLessThanOrEqual(37);
    expect(approximateCoffeeGramsForRapido(275, "Americano")).toBeGreaterThanOrEqual(15);
    expect(approximateCoffeeGramsForRapido(275, "Americano")).toBeLessThanOrEqual(17);
    expect(approximateCoffeeGramsForRapido(180, "Filter")).toBeGreaterThanOrEqual(11);
    expect(approximateCoffeeGramsForRapido(180, "Filter")).toBeLessThanOrEqual(13);
  });

  it("Rápido: approximate caffeine by type and size", () => {
    const espresso30 = estimateCaffeineMg({
      source: "brewlab",
      methodOrPreparation: "Rápido",
      coffeeGrams: 0,
      hasCaffeine: true,
      amountMl: 30,
      drinkType: "Espresso"
    });
    expect(espresso30).toBeGreaterThanOrEqual(60);
    expect(espresso30).toBeLessThanOrEqual(70);
    const mediano = estimateCaffeineMg({
      source: "brewlab",
      methodOrPreparation: "Rápido",
      coffeeGrams: 0,
      hasCaffeine: true,
      amountMl: 275,
      drinkType: "Americano"
    });
    expect(mediano).toBeGreaterThanOrEqual(140);
    expect(mediano).toBeLessThanOrEqual(150);
    const decaf = estimateCaffeineMg({
      source: "brewlab",
      methodOrPreparation: "Rápido",
      coffeeGrams: 0,
      hasCaffeine: false,
      amountMl: 180,
      drinkType: "Filter"
    });
    expect(decaf).toBeGreaterThanOrEqual(0);
    expect(decaf).toBeLessThanOrEqual(10);
  });

  it("returns dynamic brewing advice by phase/time", () => {
    const advice = getBrewingProcessAdvice("Hario V60", 16, 300, "Bloom", 4);
    expect(advice).toContain("Asegura saturacion completa");
    expect(advice).toContain("4 s");
  });
});
